compiler/rustc_mir/src/transform/separate_const_switch.rs

```diff
@@ -0,0 +1,279 @@
+            // predecessor, but for now we ignore that
+            if predecessors[block_id].len() < 2 {
+                continue 'block_iter;
+            }
+
+            // first, let's find a non-const place
+            // that determines the result of the switch
+            if let Some(switch_place) = find_determining_place(switch_place, block) {
+                // we now have an input place for which it would
+                // be interesting if predecessors assigned it from a const
+
+                let mut predecessors_left = predecessors[block_id].len();
+                'predec_iter: for predecessor_id in predecessors[block_id].iter().copied() {
+                    if let Some(predecessor) = body.basic_blocks().get(predecessor_id) {
+                        // first we make sure the predecessor jumps
+                        // in a reasonable way
+                        match &predecessor.terminator().kind {
+                            // the following terminators are
+                            // unconditionally valid
+                            TerminatorKind::Goto { .. } | TerminatorKind::SwitchInt { .. } => {}
+
+                            TerminatorKind::FalseEdge { real_target, .. } => {
+                                if *real_target != block_id {
+                                    continue 'predec_iter;
+                                }
+                            }
+
+                            // the following terminators are not allowed
+                            TerminatorKind::Resume
+                            | TerminatorKind::Drop { .. }
+                            | TerminatorKind::DropAndReplace { .. }
+                            | TerminatorKind::Call { .. }
+                            | TerminatorKind::Assert { .. }
+                            | TerminatorKind::FalseUnwind { .. }
+                            | TerminatorKind::Yield { .. }
+                            | TerminatorKind::Abort
+                            | TerminatorKind::Return
+                            | TerminatorKind::Unreachable
+                            | TerminatorKind::InlineAsm { .. }
+                            | TerminatorKind::GeneratorDrop => {
+                                continue 'predec_iter;
+                            }
+                        }
+
+                        if is_likely_const(switch_place, predecessor) {
+                            new_edges.push((predecessor_id, block_id));
+                            predecessors_left -= 1;
+                            if predecessors_left < 2 {
+                                // there is no point in duplicating anymore
+                                break 'predec_iter;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    let blocks = body.basic_blocks_mut();
+    for (pred_id, target_id) in new_edges {
```

Comment:
Can the two loops be merged?

Reply:
The reason I separated the loops is that according to somebody on Zulip, mutating the blocks vector invalidates the predecessor cache which is apparently expensive to re-compute. So this way I can have an immutable loop with a stable copy of the predecessors without having to clone it. I don't know if it's worth it. The other approach would be to clone the predecessors beforehand so we don't use the recomputed one.

Reply:
I could not come up with a good way to merge the loops. If you have a suggestion, I'd be happy to reconsider.